### PR TITLE
python311Packages.qpsolvers: specify optional dependencies

### PR DESCRIPTION
**Pull request by dotlambda**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by ghost — APPROVED** [review]
thanks. the optional-dependencies bit seems more convoluted and copy / paste than it needs to be. before i closed my change i was looking to take an array of solvers, eg: `open_source_solvers` and generate the key / value pairs but needed more time to figure out how to generate an attribute set eg:

```nix
builtins.listToAttrs  (map (f: { name = f.pname; value = f; }) (with python3Packages; [ clarabel cvxopt ]))
```

but that just looked convoluted and using `f.pname` seemed wrong.

anyway, this change seems fine.

File: pkgs/development/python-modules/qpsolvers/default.nix [hunk 1]
```diff
@@ -27,19 +32,34 @@ buildPythonPackage rec {
...
+      clarabel cvxopt daqp ecos /* highs */ osqp /* piqp proxqp qpalm */ quadprog scs
+    ];
+  };
+
   nativeCheckInputs = [
     flit-core
-    quadprog
     unittestCheckHook
-  ];
+  ] ++ passthru.optional-dependencies.open_source_solvers;
```
Comment (a-n-n-a-l-e-e) [inline pkgs/development/python-modules/qpsolvers/default.nix:62]: nit: `flit-core` above should be part of `nativeBuildInputs` not CheckInputs

**Comment by dotlambda** [comment]
> thanks. the optional-dependencies bit seems more convoluted and copy / paste than it needs to be. before i closed my change i was looking to take an array of solvers, eg: `open_source_solvers` and generate the key / value pairs but needed more time to figure out how to generate an attribute set eg:
> 
> ```nix
> builtins.listToAttrs  (map (f: { name = f.pname; value = f; }) (with python3Packages; [ clarabel cvxopt ]))
> ```
> 
> but that just looked convoluted and using `f.pname` seemed wrong.

That wouldn't work because the keys in the optional-dependencies attrset don't all correspond to the pnames.

**Comment by a-n-n-a-l-e-e** [comment]
> > thanks. the optional-dependencies bit seems more convoluted and copy / paste than it needs to be. before i closed my change i was looking to take an array of solvers, eg: `open_source_solvers` and generate the key / value pairs but needed more time to figure out how to generate an attribute set eg:
> > ```nix
> > builtins.listToAttrs  (map (f: { name = f.pname; value = f; }) (with python3Packages; [ clarabel cvxopt ]))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but that just looked convoluted and using `f.pname` seemed wrong.
> 
> That wouldn't work because the keys in the optional-dependencies attrset don't all correspond to the pnames.

they should. i think the one that doesn't is gurobipy and that is named wrong and should be renamed gurobi. also, if it is gurobipy in nixpkgs python it is not clear to me that it is correct to use `gurobi = [ gurobipy ]` in the optional-dependencies.  then again i stepped away from this change because convention around the optional-dependencies seems  ambiguous (to me).

**Review by ghost — APPROVED** [review]

**Comment by dotlambda** [comment]
The convention is to name everything exactly like upstream: https://github.com/qpsolvers/qpsolvers/blob/65050939b6f32cf2e86c5bf9bcc3019427809d5c/pyproject.toml#L35-L49

**Comment by a-n-n-a-l-e-e** [comment]
> The convention is to name everything exactly like upstream: https://github.com/qpsolvers/qpsolvers/blob/65050939b6f32cf2e86c5bf9bcc3019427809d5c/pyproject.toml#L35-L49

i guess that makes it non-ambiguous. oh and gurobi python package is gurobipy (so correct in nixpkgs) and they alias it -- weird.

**Comment by a-n-n-a-l-e-e** [comment]
aarch64-darwin is [failing](https://logs.ofborg.org/?attempt_id=451cc2e5-582f-43d1-adbd-8b14f7c837dc&key=nixos%2Fnixpkgs.287076) at osqp.

**Comment by a-n-n-a-l-e-e** [comment]
@ofborg build python311Packages.qpsolvers

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 287076` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python311Packages.qpsolvers</li>
    <li>python311Packages.qpsolvers.dist</li>
  </ul>
</details>

**Comment by a-n-n-a-l-e-e** [comment]
filed issue https://github.com/qpsolvers/qpsolvers/issues/287

**Pull request merged**